File: snapkit-js/src/voronoi.ts
/**
 * A₂ lattice Voronoï cell nearest-neighbor snap.
 *
 * Guarantees covering radius ≤ 1/√3 by checking 9 candidates
 * around the naive round-off point.
 */

// Precomputed constants
const SQRT3 = Math.sqrt(3);
const INV_SQRT3 = 1.0 / SQRT3;
const HALF_SQRT3 = 0.5 * SQRT3;

/**
 * Convert Eisenstein coordinates (a, b) to Cartesian (x, y).
 */
export function eisensteinToReal(a: number, b: number): [number, number] {
  return [a - b * 0.5, b * HALF_SQRT3];
}

/**
 * Euclidean distance from (x, y) to Eisenstein integer (a, b).
 */
export function snapDistance(x: number, y: number, a: number, b: number): number {
  const dx = x - (a - b * 0.5);
  const dy = y - (b * HALF_SQRT3);
  return Math.sqrt(dx * dx + dy * dy);
}

/**
 * Naive snap — simple rounding, no Voronoï guarantee.
 */
export function eisensteinSnapNaive(x: number, y: number): [number, number] {
  const b = Math.round(y * 2.0 * INV_SQRT3);
  const a = Math.round(x + b * 0.5);
  return [a, b];
}

/**
 * Snap (x, y) to the true nearest Eisenstein integer.
 *
 * Checks a 3×3 neighbourhood of candidates around the naive
 * round-off point using squared-distance comparison (no sqrt).
 * Tie-break by lexicographic (|a|, |b|).
 */
export function eisensteinSnapVoronoi(x: number, y: number): [number, number] {
  const b0 = Math.round(y * 2.0 * INV_SQRT3);
  const a0 = Math.round(x + b0 * 0.5);

  let bestDistSq = Infinity;
  let bestA = a0;
  let bestB = b0;

  for (let da = -1; da <= 1; da++) {
    for (let db = -1; db <= 1; db++) {
      const a = a0 + da;
      const b = b0 + db;
      const dx = x - (a - b * 0.5);
      const dy = y - (b * HALF_SQRT3);
      const dSq = dx * dx + dy * dy;
      if (dSq < bestDistSq - 1e-24) {
        bestDistSq = dSq;
        bestA = a;
        bestB = b;
      } else if (Math.abs(dSq - bestDistSq) < 1e-24) {
        // Tie-break: prefer smaller |a|, then |b|
        if (Math.abs(a) < Math.abs(bestA) || (Math.abs(a) === Math.abs(bestA) && Math.abs(b) < Math.abs(bestB))) {
          bestA = a;
          bestB = b;
        }
      }
    }
  }

  return [bestA, bestB];
}

/**
 * Batch Voronoï snap for multiple points.
 */
export function eisensteinSnapBatch(points: Array<[number, number]>): Array<[number, number]> {
  return points.map(([x, y]) => eisensteinSnapVoronoi(x, y));
}

File: snapkit-rs/src/voronoi.rs
//! 9-candidate Voronoï snap with covering radius guarantee.
//!
//! The A₂ lattice Voronoï cell contains at most 9 Eisenstein integers.
//! By testing all 9 candidates we guarantee the nearest neighbor is found,
//! and the maximum snap distance (covering radius) is ≤ 1/√3.

use crate::eisenstein::{fabs, hypot, round, EisensteinInt, HALF_SQRT3, INV_SQRT3};

/// Snap (x, y) to the true nearest Eisenstein integer using 9-candidate Voronoï search.
///
/// Guarantees covering radius ≤ 1/√3. Uses squared-distance comparison
/// (no sqrt in the inner loop) for performance.
pub fn eisenstein_round_voronoi(x: f64, y: f64) -> EisensteinInt {
    let b0 = round(y * 2.0 * INV_SQRT3) as i64;
    let a0 = round(x + b0 as f64 * 0.5) as i64;

    let mut best_dist_sq = f64::MAX;
    let mut best_a = a0;
    let mut best_b = b0;

    for da in -1..=1i64 {
        for db in -1..=1i64 {
            let a = a0 + da;
            let b = b0 + db;
            let dx = x - (a as f64 - b as f64 * 0.5);
            let dy = y - (b as f64 * HALF_SQRT3);
            let d_sq = dx * dx + dy * dy;
            if d_sq < best_dist_sq - 1e-24 {
                best_dist_sq = d_sq;
                best_a = a;
                best_b = b;
            } else if fabs(d_sq - best_dist_sq) < 1e-24 {
                // Tie-break: prefer smaller (|a|, |b|)
                if (a.abs(), b.abs()) < (best_a.abs(), best_b.abs()) {
                    best_a = a;
                    best_b = b;
                }
            }
        }
    }

    EisensteinInt::new(best_a, best_b)
}

/// Snap with distance and tolerance check using Voronoï method.
///
/// Returns `(nearest, distance, is_snap)`.
pub fn eisenstein_snap_voronoi(x: f64, y: f64, tolerance: f64) -> (EisensteinInt, f64, bool) {
    let nearest = eisenstein_round_voronoi(x, y);
    let (cx, cy) = nearest.to_cartesian();
    let dist = hypot(x - cx, y - cy);
    let is_snap = dist <= tolerance;
    (nearest, dist, is_snap)
}

/// Batch Voronoï snap for multiple points.
pub fn eisenstein_snap_voronoi_batch(points: &[(f64, f64)]) -> alloc::vec::Vec<EisensteinInt> {
    points
        .iter()
        .map(|&(x, y)| eisenstein_round_voronoi(x, y))
        .collect()
}

/// Batch Voronoï snap with distance and tolerance check.
pub fn eisenstein_snap_voronoi_batch_with_tolerance(
    points: &[(f64, f64)],
    tolerance: f64,
) -> alloc::vec::Vec<(EisensteinInt, f64, bool)> {
    points
        .iter()
        .map(|&(x, y)| eisenstein_snap_voronoi(x, y, tolerance))
        .collect()
}

/// Verify the covering radius guarantee: maximum snap distance ≤ 1/√3.
///
/// Tests a dense grid of points and returns the maximum observed snap distance.
pub fn verify_covering_radius(grid_resolution: usize) -> f64 {
    let mut max_dist = 0.0_f64;
    let step = 1.0 / grid_resolution as f64;
    // Test points in [0, 1]² — one Voronoï cell
    let n = grid_resolution;
    for i in 0..=n {
        for j in 0..=n {
            let x = i as f64 * step;
            let y = j as f64 * step;
            let nearest = eisenstein_round_voronoi(x, y);
            let (cx, cy) = nearest.to_cartesian();
            let dist = hypot(x - cx, y - cy);
            if dist > max_dist {
                max_dist = dist;
            }
        }
    }
    max_dist
}

extern crate alloc;
